packet/packet.go

```diff
@@ -213,6 +226,53 @@ func (packet *Packet) ParseL3() {
 	packet.L3 = unsafe.Pointer(packet.unparsed())
 }
 
+// Correctly returns EtherType from Ethernet header or VLAN header.
+func (packet *Packet) GetEtherType() uint16 {
+	if packet.Ether.EtherType == SwapBytesUint16(VLANNumber) {
+		vptr := uintptr(unsafe.Pointer(packet.Ether)) + EtherLen
+		vhdr := (*VLANHdr)(unsafe.Pointer(vptr))
+		return vhdr.EtherType
+	} else {
+		return packet.Ether.EtherType
+	}
+}
+
+// ParseL3 set poinetr to start of L3 header taking possible presence
+// of VLAN header into account.
+func (packet *Packet) ParseL3CheckVLAN() {
+	ptr := uintptr(unsafe.Pointer(packet.Ether)) + EtherLen
+
+	if packet.Ether.EtherType == SwapBytesUint16(VLANNumber) {
+		ptr += VLANLen
+	}
+	packet.L3 = unsafe.Pointer(ptr)
+}
+
+// ParseL3 set poinetr to start of L3 header taking possible VLAN
+// presence into account. It returns VLAN header if present and always
+// correct EtherType from VLAN header or Ethernet header.
+func (packet *Packet) ParseVLANAndL3() (uint16, *VLANHdr) {
```

Comment:
I would like to name this function ParseL3ReturnVLAN. Also I suppose that this functions should return only vhdr. User may compare it to to nil after this function.